frontends/common/parser_options.h

```diff
@@ -54,6 +52,33 @@ class ParserOptions : public Util::Options {
 
     std::vector<const char *> *process(int argc, char *const argv[]) override;
     enum class FrontendVersion { P4_14, P4_16 };
+
+    /// Records the result of the preprocessor.
+    class PreprocessorResult {
+     private:
+        /// The input stream.
+        FILE *_file = nullptr;
+        /// Whether the input stream should be closed.
+        bool _closeInput = false;
+
+        /// Tries to close the input stream associated with the result.
+        void closeFile();
+
+     public:
+        PreprocessorResult() = default;
+        PreprocessorResult &operator=(PreprocessorResult &&) = default;
+        PreprocessorResult(PreprocessorResult &&) = default;
```

Comment:
This will not work. The default move will do memberwise move, but for primitive types that is just a copy. So we either need to disable the move constructor too (which would probably work, otherwise there is a double-close somewhere now), or explicitly set `orig._file = nullptr` in the move operator/ctor.

Reply:
Or, maybe, one could use `std::unique_ptr<FILE, decltype(&fclose)>` This way the owning semantics would be preserved.